File: build-data.js
#!/usr/bin/env node
/**
 * Build script: Converts mymind CSV export + resolved TikTok URLs into data.js
 * Run once: node build-data.js
 */

const fs = require('fs');
const path = require('path');

// Simple CSV parser that handles quoted fields, commas in values, and multiline
function parseCSV(text) {
    const rows = [];
    let current = '';
    let inQuotes = false;
    const lines = text.split('\n');

    for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        if (inQuotes) {
            current += '\n' + line;
            if (line.includes('"')) {
                // Check if quote is closing
                let quoteCount = 0;
                for (const ch of line) {
                    if (ch === '"') quoteCount++;
                }
                if (quoteCount % 2 === 1) {
                    inQuotes = false;
                    rows.push(current);
                    current = '';
                }
            }
        } else {
            // Count quotes in line
            let quoteCount = 0;
            for (const ch of line) {
                if (ch === '"') quoteCount++;
            }
            if (quoteCount % 2 === 1) {
                inQuotes = true;
                current = line;
            } else {
                rows.push(line);
            }
        }
    }

    // Parse each row into fields
    return rows.filter(r => r.trim()).map(row => {
        const fields = [];
        let field = '';
        let inQ = false;

        for (let i = 0; i < row.length; i++) {
            const ch = row[i];
            if (ch === '"') {
                if (inQ && row[i + 1] === '"') {
                    field += '"';
                    i++;
                } else {
                    inQ = !inQ;
                }
            } else if (ch === ',' && !inQ) {
                fields.push(field);
                field = '';
            } else {
                field += ch;
            }
        }
        fields.push(field);
        return fields;
    });
}

// Extract YouTube video ID from URL
function extractYouTubeId(url) {
    const match = url.match(/(?:(?:m\.)?youtube\.com\/(?:watch\?(?:.*&)?v=|embed\/|shorts\/|live\/)|youtu\.be\/)([a-zA-Z0-9_-]{11})/i);
    return match ? match[1] : null;
}

// Extract Instagram post ID from URL
function extractInstagramId(url) {
    const match = url.match(/instagram\.com\/(?:reel|p|tv)\/([a-zA-Z0-9_-]+)/i);
    return match ? match[1] : null;
}

// Extract Twitter/X tweet ID from URL
function extractTwitterId(url) {
    const match = url.match(/(?:twitter\.com|x\.com)\/(?:#!\/)?(\w+)\/status(?:es)?\/(\d+)/i);
    return match ? { user: match[1], id: match[2] } : null;
}

// Get file extension for local media
function getLocalFileExt(id, mediaDir) {
    const exts = ['.jpeg', '.jpg', '.png', '.webp', '.gif', '.mov', '.mp4', '.pdf'];
    for (const ext of exts) {
        const subDirs = ['images', 'videos', 'docs'];
        for (const sub of subDirs) {
            if (fs.existsSync(path.join(mediaDir, sub, id + ext))) {
                return { ext, subDir: sub, filename: id + ext };
            }
        }
    }
    return null;
}

function main() {
    // Read CSV
    const csvText = fs.readFileSync('cards.csv', 'utf-8').replace(/^\uFEFF/, ''); // Remove BOM
    const rows = parseCSV(csvText);

    if (rows.length === 0) {
        console.error('No rows found in CSV');
        process.exit(1);
    }

    // First row is headers
    const headers = rows[0];
    console.log('Headers:', headers);

    // Read resolved TikTok URLs
    let tiktokResolved = {};
    try {
        tiktokResolved = JSON.parse(fs.readFileSync('tiktok-resolved.json', 'utf-8'));
    } catch (e) {
        console.warn('No tiktok-resolved.json found, TikTok embeds will be link-only');
    }

    const mediaDir = path.join(__dirname, 'media');
    const entries = [];
    let skipped = 0;

    for (let i = 1; i < rows.length; i++) {
        const fields = rows[i];
        if (fields.length < 8) {
            skipped++;
            continue;
        }

        const [id, type, title, url, content, note, tags, created] = fields;

        if (!id || !type) {
            skipped++;
            continue;
        }

        const entry = {
            id: id.trim(),
            title: (title || '').trim(),
            url: (url || '').trim(),
            content: (content || '').trim(),
            note: (note || '').trim(),
            tags: (tags || '').split(',').map(t => t.trim()).filter(Boolean),
            timestamp: created ? new Date(created.trim()).getTime() : Date.now(),
            source: 'mymind',
        };

        // Map CSV type to app type
        switch (type.trim()) {
            case 'InstagramReel':
            case 'InstagramPost': {
                entry.platform = 'Instagram';
                const igId = extractInstagramId(entry.url);
                if (igId) {
                    entry.embedType = 'iframe';
                    entry.embedUrl = `https://www.instagram.com/p/${igId}/embed/?captioned=false`;
                } else {
                    entry.embedType = 'link';
                }
                break;
            }

            case 'TikTokPost':
            case 'TikTok': {
                entry.platform = 'TikTok';
                const resolved = tiktokResolved[entry.id];
                if (resolved && resolved.video_id) {
                    entry.embedType = 'iframe';
                    entry.embedUrl = `https://www.tiktok.com/embed/v2/${resolved.video_id}`;
                    entry.resolvedUrl = resolved.full_url;
                } else {
                    entry.embedType = 'link';
                }
                break;
            }

            case 'YouTubeVideo': {
                entry.platform = 'YouTube';
                const ytId = extractYouTubeId(entry.url);
                if (ytId) {
                    entry.embedType = 'iframe';
                    entry.embedUrl = `https://www.youtube.com/embed/${ytId}?rel=0&modestbranding=1`;
                } else {
                    entry.embedType = 'link';
                }
                break;
            }

            case 'XPost': {
                entry.platform = 'Twitter/X';
                const tweet = extractTwitterId(entry.url);
                if (tweet) {
                    entry.embedType = 'iframe';
                    entry.embedUrl = `https://platform.twitter.com/embed/Tweet.html?id=${tweet.id}&theme=dark&hide_thread=true`;
                } else {
                    entry.embedType = 'link';
                }
                break;
            }

            case 'Image': {
                entry.platform = 'Image';
                // Check for local file
                const localFile = getLocalFileExt(entry.id, mediaDir);
                if (localFile) {
                    entry.embedType = 'image';
                    entry.mediaFile = `media/${localFile.subDir}/${localFile.filename}`;
                } else if (entry.url) {
                    entry.embedType = 'image';
                    entry.mediaFile = entry.url;
                } else {
                    entry.embedType = 'text';
                }
                break;
            }

            case 'Video': {
                entry.platform = 'Video';
                const localVid = getLocalFileExt(entry.id, mediaDir);
                if (localVid) {
                    entry.embedType = 'video';
                    entry.mediaFile = `media/${localVid.subDir}/${localVid.filename}`;
                } else {
                    entry.embedType = 'link';
                }
                break;
            }

            case 'Document': {
                entry.platform = 'Document';
                const localDoc = getLocalFileExt(entry.id, mediaDir);
                if (localDoc) {
                    entry.embedType = 'document';
                    entry.mediaFile = `media/${localDoc.subDir}/${localDoc.filename}`;
                } else {
                    entry.embedType = 'text';
                }
                break;
            }

            case 'WebPage': {
                entry.platform = 'Web';
                entry.embedType = 'link';
                break;
            }

            case 'Article': {
                entry.platform = 'Article';
                entry.embedType = 'link';
                break;
            }

            case 'Note':
            case 'Post':
            case 'Content': {
                entry.platform = type.trim();
                entry.embedType = 'text';
                break;
            }

            default: {
                entry.platform = 'Other';
                // Check if it has a URL
                if (entry.url) {
                    entry.embedType = 'link';
                } else {
                    entry.embedType = 'text';
                }
                break;
            }
        }

        entries.push(entry);
    }

    // Sort by timestamp (newest first)
    entries.sort((a, b) => b.timestamp - a.timestamp);

    // Generate data.js
    const output = `// Auto-generated from mymind CSV export
// Generated: ${new Date().toISOString()}
// Total entries: ${entries.length}
window.MYMIND_DATA = ${JSON.stringify(entries, null, 2)};
`;

    fs.writeFileSync('data.js', output, 'utf-8');

    // Stats
    const stats = {};
    entries.forEach(e => {
        const key = `${e.platform}/${e.embedType}`;
        stats[key] = (stats[key] || 0) + 1;
    });

    console.log(`\nGenerated data.js with ${entries.length} entries (${skipped} skipped)`);
    console.log('\nBreakdown:');
    Object.entries(stats).sort((a, b) => b[1] - a[1]).forEach(([key, count]) => {
        console.log(`  ${count.toString().padStart(4)} ${key}`);
    });
}

main();

File: build-data.py
#!/usr/bin/env python3
"""
Build script: Converts mymind CSV export + resolved TikTok URLs into data.js
Run once: python3 build-data.py
"""

import csv
import json
import os
import re
from datetime import datetime

def extract_youtube_id(url):
    match = re.search(
        r'(?:(?:m\.)?youtube\.com\/(?:watch\?(?:.*&)?v=|embed\/|shorts\/|live\/)|youtu\.be\/)([a-zA-Z0-9_-]{11})',
        url, re.I
    )
    return match.group(1) if match else None

def extract_instagram_id(url):
    match = re.search(r'instagram\.com\/(?:reel|p|tv)\/([a-zA-Z0-9_-]+)', url, re.I)
    return match.group(1) if match else None

def extract_twitter_id(url):
    match = re.search(r'(?:twitter\.com|x\.com)\/(?:#!\/)?(\w+)\/status(?:es)?\/(\d+)', url, re.I)
    return {'user': match.group(1), 'id': match.group(2)} if match else None

def get_local_file(entry_id, media_dir):
    exts = ['.jpeg', '.jpg', '.png', '.webp', '.gif', '.mov', '.mp4', '.pdf']
    sub_dirs = {'images': ['.jpeg', '.jpg', '.png', '.webp', '.gif'],
                'videos': ['.mov', '.mp4'],
                'docs': ['.pdf']}
    for sub, valid_exts in sub_dirs.items():
        for ext in valid_exts:
            filepath = os.path.join(media_dir, sub, entry_id + ext)
            if os.path.exists(filepath):
                return f'media/{sub}/{entry_id}{ext}'
    return None

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    csv_path = os.path.join(script_dir, 'cards.csv')
    tiktok_path = os.path.join(script_dir, 'tiktok-resolved.json')
    media_dir = os.path.join(script_dir, 'media')

    # Read resolved TikTok URLs
    tiktok_resolved = {}
    try:
        with open(tiktok_path, 'r') as f:
            tiktok_resolved = json.load(f)
    except FileNotFoundError:
        print('Warning: No tiktok-resolved.json found')

    # Read CSV
    entries = []
    skipped = 0

    with open(csv_path, 'r', encoding='utf-8-sig') as f:
        reader = csv.DictReader(f)
        for row in reader:
            entry_id = (row.get('id') or '').strip()
            entry_type = (row.get('type') or '').strip()
            title = (row.get('title') or '').strip()
            url = (row.get('url') or '').strip()
            content = (row.get('content') or '').strip()
            note = (row.get('note') or '').strip()
            tags_str = (row.get('tags') or '').strip()
            created = (row.get('created') or '').strip()

            if not entry_id or not entry_type:
                skipped += 1
                continue

            # Parse tags
            tags = [t.strip() for t in tags_str.split(',') if t.strip()]

            # Parse timestamp
            try:
                ts = int(datetime.fromisoformat(created.replace('Z', '+00:00')).timestamp() * 1000)
            except (ValueError, AttributeError):
                ts = int(datetime.now().timestamp() * 1000)

            entry = {
                'id': entry_id,
                'title': title,
                'url': url,
                'tags': tags,
                'timestamp': ts,
                'source': 'mymind',
            }

            # Only include content/note if non-empty
            if content:
                entry['content'] = content
            if note:
                entry['note'] = note

            # Map CSV type to app type
            if entry_type in ('InstagramReel', 'InstagramPost'):
                entry['platform'] = 'Instagram'
                ig_id = extract_instagram_id(url) if url else None
                if ig_id:
                    entry['embedType'] = 'iframe'
                    entry['embedUrl'] = f'https://www.instagram.com/p/{ig_id}/embed/?captioned=false'
                else:
                    entry['embedType'] = 'link'

            elif entry_type in ('TikTokPost', 'TikTok'):
                entry['platform'] = 'TikTok'
                resolved = tiktok_resolved.get(entry_id)
                if resolved and resolved.get('video_id'):
                    entry['embedType'] = 'iframe'
                    entry['embedUrl'] = f'https://www.tiktok.com/embed/v2/{resolved["video_id"]}'
                else:
                    entry['embedType'] = 'link'

            elif entry_type == 'YouTubeVideo':
                entry['platform'] = 'YouTube'
                yt_id = extract_youtube_id(url) if url else None
                if yt_id:
                    entry['embedType'] = 'iframe'
                    entry['embedUrl'] = f'https://www.youtube.com/embed/{yt_id}?rel=0&modestbranding=1'
                else:
                    entry['embedType'] = 'link'

            elif entry_type == 'XPost':
                entry['platform'] = 'Twitter/X'
                tweet = extract_twitter_id(url) if url else None
                if tweet:
                    entry['embedType'] = 'iframe'
                    entry['embedUrl'] = f'https://platform.twitter.com/embed/Tweet.html?id={tweet["id"]}&theme=dark&hide_thread=true'
                else:
                    entry['embedType'] = 'link'

            elif entry_type == 'Image':
                entry['platform'] = 'Image'
                local_file = get_local_file(entry_id, media_dir)
                if local_file:
                    entry['embedType'] = 'image'
                    entry['mediaFile'] = local_file
                elif url:
                    entry['embedType'] = 'image'
                    entry['mediaFile'] = url
                else:
                    entry['embedType'] = 'text'

            elif entry_type == 'Video':
                entry['platform'] = 'Video'
                local_file = get_local_file(entry_id, media_dir)
                if local_file:
                    entry['embedType'] = 'video'
                    entry['mediaFile'] = local_file
                else:
                    entry['embedType'] = 'link'

            elif entry_type == 'Document':
                entry['platform'] = 'Document'
                local_file = get_local_file(entry_id, media_dir)
                if local_file:
                    entry['embedType'] = 'document'
                    entry['mediaFile'] = local_file
                else:
                    entry['embedType'] = 'text'

            elif entry_type == 'WebPage':
                entry['platform'] = 'Web'
                entry['embedType'] = 'link'

            elif entry_type == 'Article':
                entry['platform'] = 'Article'
                entry['embedType'] = 'link'

            elif entry_type in ('Note', 'Post', 'Content'):
                entry['platform'] = entry_type
                entry['embedType'] = 'text'

            else:
                entry['platform'] = 'Other'
                entry['embedType'] = 'link' if url else 'text'

            entries.append(entry)

    # Sort by timestamp (newest first)
    entries.sort(key=lambda e: e['timestamp'], reverse=True)

    # Generate data.js
    output = f'''// Auto-generated from mymind CSV export
// Generated: {datetime.now().isoformat()}
// Total entries: {len(entries)}
window.MYMIND_DATA = {json.dumps(entries, indent=2)};
'''

    output_path = os.path.join(script_dir, 'data.js')
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(output)

    # Stats
    stats = {}
    for e in entries:
        key = f"{e['platform']}/{e['embedType']}"
        stats[key] = stats.get(key, 0) + 1

    print(f'Generated data.js with {len(entries)} entries ({skipped} skipped)')
    print(f'\nBreakdown:')
    for key, count in sorted(stats.items(), key=lambda x: -x[1]):
        print(f'  {count:4d} {key}')

if __name__ == '__main__':
    main()
